### DISCUSSION TITLE
Please help with an example of the client.

### DISCUSSION THREAD

**Discussion by neptune-v** [modelcontextprotocol/modelcontextprotocol#1206] [Q&A - Client implementation] [answered]
In the official docs, all the examples assume the client runs inside the Claude Desktop App alike (host).
What if I am building my app (host) and want to adopt MCP?
I have difficulty understanding the flow between the LLM API (e.g., the Anthropic API) and the MCP client inside the host.
Could you point me to some client examples that can demonstrate this part of the flow?

**Comment by RyoJerryYu** [comment]
I've encountered the same difficulty myself. I've been searching for examples of MCP Client implementations, but unfortunately, the only resource I've found is this very discussion.

It would be incredibly helpful to have a sample implementation of an MCP Client, even if it's just a simple command-line tool.

**Reply by dsp-ant** [reply]
We totally understand where you are coming from. There is certainly a lack of a good client guide and implementation at the moment. We are working on it. For now, feel free to start discussions if you encounter issues and we are happy to help to the best of our ability.

**Reply by neptune-v** [reply]
Thanks for the reply, and I look forward to studying the client reference implementation.
Please let me know if I am wrong. I understand that MCP only acts as the abstraction and segregation of concern, so the MCP server handles the actual function execution. The MCP client reads the available functions and acts as the delegate to call the functions on the server.
So, I guess the Anthropic client still uses "tools" but discovers available tools via the MCP client fetched from the MCP server. Instead of calling a pre-defined function locally, the Anthropic client asks the MCP client to delegate the function call to the MCP server.
This way:
1. The available functions can be developed according to a unified MCP.
2. Any MCP app can easily reuse these functions.
3. To some degree, The delegated function call and the MCP server prevent the Anthropic API from being tricked into executing dangerous code locally.

**Reply by rockywangxiaolei** [reply]
Purely speculative:
From the way Claude Desktop uses the MCP Server (via config.json and logs), it seems that the client is responsible for reading the list of MCP server tools in the configuration, then establishing a client-server connection protocol. Once confirmed, the client uses the corresponding tools based on user input. In the stdio mode, the client achieves this by launching a local runtime environment (such as Node.js or UVX) to run the compiled code from the MCP server.

If SSE and a remote MCP server are adopted, I speculate that the client would only need to pass the tools and parameters to the MCP server, leaving the MCP server to execute the tasks based on triggers.

**Reply by solarslurpi** [reply]
I want to write a client. Right now, it is "painful" to use the Claude desktop because quickly I get the message to subscribe for $20 to get 5 more somethings.  But then what happens when I go over that doing these mcp client/server things? Which it was more like other open protocols like mqtt.

**Comment by AaronAbuUsama** [comment]
i will take a crack at it.

**Comment by ArsenicBismuth** [comment, marked as answer]
I wrote one before, in the `python-sdk` repo. Unfortunately they removed the discussion section alongside my writeup.

EDIT: Nevermind, it's moved into this discussion. You can check my implementation here: https://github.com/orgs/modelcontextprotocol/discussions/36

**Reply by neptune-v** [reply]
Thanks for your example, sir! @dsp-ant 's comment also confirms my observation. I am marking this as the answer.
https://github.com/orgs/modelcontextprotocol/discussions/36#discussioncomment-11389347
Still looking forward to seeing more comprehensive official examples in the future!

**Reply by ArsenicBismuth** [reply]
@neptune-v Yeah, I was also hoping for that. Coz I'm also not sure the best practice around making the client (such as when is the server context is necessary or not).

I updated my top-level discussion with the full MVP that I use for clarity.

**Reply by solarslurpi** [reply]
Thank you @ArsenicBismuth .  Off I go to see your write up. Thank you for making us aware of your work.

**Comment by adhikasp** [comment]
Hey I wrote a simple Python CLI client for where the LLM can do tool call to MCP servers, maybe useful for your references too

The main bit of logic
https://github.com/adhikasp/mcp-client-cli/blob/master/src/mcp_client_cli/cli.py#L187-L221

**Comment by localden** [comment]
We've limited Discussions in this repository to meeting notes and are closing this thread. To continue the conversation, pick a channel from [Contributor Communication](https://modelcontextprotocol.io/community/communication). Thanks for posting.
